### PR TITLE
docs: fix bug in code examples for downgrading a service

### PR DESCRIPTION
**Pull request by bertyhell**

### CODE REVIEW

**Comment by gkalpak** [comment]
The service is indeed called `heroes` in [AngularJS-land][1]. This change seems invalid.

[1]: https://github.com/bertyhell/angular/blob/90d0f79b9fed1435e6c97ea05fdcb62d33aea570/aio/content/examples/upgrade-module/src/app/ajs-to-a-providers/app.module.ts#L38

**Comment by bertyhell** [comment]
> The service is indeed called heroes in AngularJS-land. This change seems invalid.

If that is the case, the documentation is even more confusing, since it states the opposite here:
https://angular.io/guide/upgrade#making-angularjs-dependencies-injectable-to-angular

> In these situations, it is possible to upgrade an AngularJS provider to Angular. This makes it possible to then inject it somewhere in Angular code. For example, you might have a service called HeroesService in AngularJS:

**Comment by gkalpak** [comment]
It is indeed somewhat confusing, because we never show how the service is registered in AngularJS. In both Angular and AngularJS there are two things that identify a registered service:
1. The injection token, i.e. an identifier that we will use to request the service from the injector.
2. The service class (which will be used by the injector to create an instance of the service).

In Angular, it is common to use the class itself as the injection token:

```ts
providers: [
  {provide: MyClass, useClass: MyClass},
  // or the shorthand:
  MyClass
]
```

This is possible, because Angular's injection system can accept any token as identifier.

In AngularJS on the other hand, we only accept string identifiers as token. Thus you are forced to use two different things: a string (for the injection token) and a class (for the service)

```ts
.service('myServiceId', MyServiceClass)
```

It is common in AngularJS apps to chose a service name for the token (e.g. `user`) and append the `Service` suffix to create the class name:

```ts
.service('user', class UserService { ... })
```

##
Hopefully that clears it up a bit. Feel free to update to clarify things (as long as the change is valid :smiley:).

**Comment by bertyhell** [comment]
Ok, I changed the token name back to the original and added a note to explain the 'heroes' string.
Does that work for you?

https://github.com/angular/angular/pull/19371/files#diff-33c8eb0a067223c3fbf3dd45ec689efdR783

**Comment by trotyl** [comment]
> This is possible, because Angular's injection system can accept any token as identifier.

Hi @gkalpak , sorry for off topic, I guess there're some different voices on https://github.com/angular/angular/pull/15319#issuecomment-287965852, may I know what is the expected validity for DI token now (or planned future)?

**Comment by bertyhell** [comment]
@trotyl Good point, the question is: After the 5.0 release will people be upgrading their app from 1.x => 4.x or 1.x => 5.x.

Either way, the docs will have to be updated to reflect this, but that's beyond the scope of this PR.

**Comment by gkalpak** [comment]
@trotyl, yes, things might change in future version. I am not sure what the current state is.
@bertyhell, what we have in the docs now will work with 4.x and 5.x alike, since we are using a `Type` (`HeroesService`) as token (in this case at least).

File: aio/content/examples/upgrade-module/src/app/ajs-to-a-providers/hero-detail.component.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ import { Hero } from '../hero';
 
 @Component({
   selector: 'hero-detail',
+  providers: [HeroesService],
```
Comment (gkalpak) [inline aio/content/examples/upgrade-module/src/app/ajs-to-a-providers/hero-detail.component.ts]: Why this change?

Comment (bertyhell) [reply aio/content/examples/upgrade-module/src/app/ajs-to-a-providers/hero-detail.component.ts]: @gkalpak Without this change I get this error in the app i'm trying to migrate:

```
ERROR Error: Uncaught (in promise): Error: [$injector:unpr] Unknown provider: heroesServiceProvider <- heroesService
```

Comment (gkalpak) [reply aio/content/examples/upgrade-module/src/app/ajs-to-a-providers/hero-detail.component.ts]: This example provides the service on the `NgModule`. Providing it on the component too, will create a separate instance for the component (different than what the rest of the app is using). This is not what we want here (but might be what you want in your app).

To share the same instance of the service across components, put it on the `providers` array of the `NgModule`.

**Comment by bertyhell** [comment]
> To share the same instance of the service across components, put it on the providers array of the NgModule.

I modified the code and added some more description text to the readme.

File: aio/content/guide/upgrade.md [hunk 2]
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
 It's also recommended to export the `heroesServiceFactory` function so that Ahead-of-Time
 compilation can pick it up.
 
+<div class="l-sub-section">
+
+Note that the 'heroes' string inside the factory refers to the AngularJS `HeroesService`.
+It is common in AngularJS apps to chose a service name for the token (e.g. heroes)
+and append the Service suffix to create the class name (e.g. heroesService)
```
Comment (kapunahelewong) [inline aio/content/guide/upgrade.md]: Could you please change "class name (e.g. heroesService)" to "class name, for example `heroesService`"? The reasons for changes are:

1. We are trying to minimize parentheticals whenever we can.
2. We're replacing i.e. and e.g. with "that is" and "for example" since people often confuse them.
3. For consistency, `heroesService` should be in code font, thus the back ticks.

File: aio/content/guide/upgrade.md [hunk 3]
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
...
 <code-example path="upgrade-module/src/app/ajs-to-a-providers/ajs-upgraded-providers.ts" title="ajs-upgraded-providers.ts">
 </code-example>
 
+You can then add the service to Angular by adding it to the `@NgModule`:
+
 <code-example path="upgrade-module/src/app/ajs-to-a-providers/app.module.ts" region="register" title="app.module.ts">
 </code-example>
 
-You can then inject it in Angular using it's class as a type annotation:
+Then use the service inside your component by injecting it in the component constructor using it's class as a type annotation:
```
Comment (kapunahelewong) [inline aio/content/guide/upgrade.md]: it's --> its

I know this was already there, but since you're on the line, would you mind deleting the apostrophe?

**Review by kapunahelewong — CHANGES_REQUESTED** [review]
Thanks for the updates @bertyhell. Just a couple of comments for you.

**Comment by bertyhell** [comment]
@kapunahelewong Sorry for the long delay. I missed the email notification about your comments.

**Comment by kapunahelewong** [comment]
Hey @bertyhell, if you'd like to rebase, I'd be happy to approve :)

**Comment by brandonroberts** [comment]
@gkalpak are there any changes left on this from your point of view?

File: aio/content/guide/upgrade.md [hunk 4]
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
 It's also recommended to export the `heroesServiceFactory` function so that Ahead-of-Time
 compilation can pick it up.
 
+<div class="l-sub-section">
```
Comment (brandonroberts) [inline aio/content/guide/upgrade.md]: Change `l-sub-section` to `alert is-helpful`

**Review by brandonroberts — CHANGES_REQUESTED** [review]
Minor css change

File: aio/content/guide/upgrade.md [hunk 5]
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
...
+
+Note that the 'heroes' string inside the factory refers to the AngularJS `HeroesService`.
+It is common in AngularJS apps to chose a service name for the token, for example heroes,
```
Comment (gkalpak) [inline aio/content/guide/upgrade.md]: _chose_ --> _choose_
_for example heroes_ --> _for example "heroes"_

File: aio/content/guide/upgrade.md [hunk 6]
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
...
+and append the Service suffix to create the class name, for example `heroesService`.
```
Comment (gkalpak) [inline aio/content/guide/upgrade.md]: _Service_ --> _"Service"_
Also, classes (or constructor functions) start with an uppercase letter (by convention). So `HeroesService`, not `heroesService`.

File: aio/content/guide/upgrade.md [hunk 7]
```diff
@@ -778,16 +778,25 @@ delete them once the upgrade is over.
...
+
+</div>
+
 <code-example path="upgrade-module/src/app/ajs-to-a-providers/ajs-upgraded-providers.ts" title="ajs-upgraded-providers.ts">
 </code-example>
 
+You can then add the service to Angular by adding it to the `@NgModule`:
```
Comment (gkalpak) [inline aio/content/guide/upgrade.md]: I would change "add the service" to "provide the service".

**Review by gkalpak — COMMENTED** [review]
a couple of minor comments. LGTM otherwise.

**Comment by brandonroberts** [comment]
@gkalpak this one is ready for another review

**Review by gkalpak — APPROVED** [review]
https://github.com/angular/angular/pull/19371#discussion_r207909524 hasn't been fully addressed. Other than that LGTM.

**Review by brandonroberts — APPROVED** [review]

**Review by kapunahelewong — APPROVED** [review]
